File: runtime-erlang/server/smithy_router.erl
-module(smithy_router).

%% Request routing for Smithy-generated Erlang servers.
%%
%% A generated server module calls `smithy_router:dispatch/4` with the
%% parsed request components and a routing table built from the service's
%% Smithy model.  Each route entry maps `{Method, PathPattern}` to a
%% handler callback `{Module, Function}`.
%%
%% Path patterns follow the Smithy URI template syntax:
%%   "/items/{id}"        → matches "/items/42",  binds Id = <<"42">>
%%   "/prefix/{+proxy}"  → greedy label, matches the rest of the path
%%
%% ## Route table format
%%
%%   Routes :: [{Method, PathPattern, Module, Function}]
%%
%%   Method      :: binary()  (uppercase, e.g. <<"GET">>)
%%   PathPattern :: binary()  (e.g. <<"/items/{id}">>)
%%   Module      :: module()
%%   Function    :: atom()
%%
%% ## Dispatch protocol
%%
%% The handler `Module:Function/3` is called as:
%%   Module:Function(PathBindings, Headers, Body)
%% where `PathBindings` is a map of `binary() => binary()` label values.
%%
%% The handler must return one of:
%%   {ok, Result}        → 200 JSON response (Result is JSON-encodable)
%%   {ok, Status, Body}  → custom status with pre-encoded body
%%   {error, Reason}     → delegated to smithy_server:error_response/1
%%   not_found           → 404

-export([
    dispatch/4,
    match/2
]).

%% Internal (exported for testing)
-export([
    match_path/2,
    extract_bindings/2,
    split_path/1
]).

-type method()       :: binary().
-type path_pattern() :: binary().
-type route()        :: {method(), path_pattern(), module(), atom()}.
-type routes()       :: [route()].
-type bindings()     :: #{binary() => binary()}.
-type headers()      :: #{binary() => binary()} | [{binary(), binary()}].
-type body()         :: binary().

%%====================================================================
%% Public API
%%====================================================================

%% @doc Dispatch an incoming request to the matching handler.
%%
%% Routes are looked up from the application environment key
%% `{smithy_beam, routes}`.  Register routes before calling this function
%% (typically in your application's `start/2` callback):
%%
%%   application:set_env(smithy_beam, routes, MyServer:routes())
%%
%% @param Method   Uppercase HTTP method binary (e.g. `<<"POST">>`).
%% @param Path     URL path binary (e.g. `<<"/items/42">>`).
%% @param Headers  Request headers map or list.
%% @param Body     Raw request body binary.
%% @returns Handler return value, `not_found`, or `{error, routes_not_configured}`.
-spec dispatch(method(), binary(), headers(), body()) -> term().
dispatch(Method, Path, Headers, Body) ->
    case application:get_env(smithy_beam, routes) of
        {ok, Routes} -> dispatch(Routes, Method, Path, Headers, Body);
        undefined    -> {error, routes_not_configured}
    end.

%% @doc Dispatch using an explicit route table.
%%
%% Iterates the route table in order, stopping at the first match.
%% Calls the matched handler with `(Bindings, Headers, Body)`.
%%
%% @param Routes   Route table (list of 4-tuples).
%% @param Method   Uppercase HTTP method binary.
%% @param Path     URL path binary.
%% @param Headers  Request headers.
%% @param Body     Raw request body binary.
%% @returns Handler return value or `not_found`.
-spec dispatch(routes(), method(), binary(), headers(), body()) -> term().
dispatch([], _Method, _Path, _Headers, _Body) ->
    not_found;
dispatch([{RouteMethod, Pattern, Mod, Fun} | Rest], Method, Path, Headers, Body) ->
    case RouteMethod =:= Method orelse RouteMethod =:= <<"*">> of
        false ->
            dispatch(Rest, Method, Path, Headers, Body);
        true ->
            case match_path(Pattern, Path) of
                {ok, Bindings} ->
                    Mod:Fun(Bindings, Headers, Body);
                nomatch ->
                    dispatch(Rest, Method, Path, Headers, Body)
            end
    end.

%% @doc Return the first matching route entry for the given method and path.
%%
%% @returns `{ok, Bindings, Module, Function}` or `nomatch`.
-spec match(routes(), {method(), binary()}) ->
    {ok, bindings(), module(), atom()} | nomatch.
match([], _) ->
    nomatch;
match([{RouteMethod, Pattern, Mod, Fun} | Rest], {Method, Path}) ->
    case RouteMethod =:= Method orelse RouteMethod =:= <<"*">> of
        false ->
            match(Rest, {Method, Path});
        true ->
            case match_path(Pattern, Path) of
                {ok, Bindings} -> {ok, Bindings, Mod, Fun};
                nomatch        -> match(Rest, {Method, Path})
            end
    end.

%%====================================================================
%% Path matching
%%====================================================================

%% @doc Test whether `Path` matches `Pattern`, returning label bindings.
%%
%% Pattern segments enclosed in `{...}` are label captures.
%% A label prefixed with `+` (i.e. `{+proxy}`) greedily captures the
%% remainder of the path including `/` separators.
%%
%% @returns `{ok, Bindings}` or `nomatch`.
-spec match_path(path_pattern(), binary()) -> {ok, bindings()} | nomatch.
match_path(Pattern, Path) ->
    PatternSegs = split_path(Pattern),
    PathSegs    = split_path(Path),
    match_segments(PatternSegs, PathSegs, #{}).

match_segments([], [], Bindings) ->
    {ok, Bindings};
match_segments([<<"{+", Rest/binary>> | _PatRest], PathSegs, Bindings) ->
    %% Greedy label: consumes all remaining path segments
    LabelName = binary:part(Rest, 0, byte_size(Rest) - 1),  %% strip trailing }
    Value = iolist_to_binary(lists:join(<<"/">>, PathSegs)),
    {ok, Bindings#{LabelName => Value}};
match_segments([<<"{"  , Rest/binary>> | PatRest], [PathSeg | PathRest], Bindings) ->
    LabelName = binary:part(Rest, 0, byte_size(Rest) - 1),
    match_segments(PatRest, PathRest, Bindings#{LabelName => PathSeg});
match_segments([Literal | PatRest], [Literal | PathRest], Bindings) ->
    match_segments(PatRest, PathRest, Bindings);
match_segments(_, _, _) ->
    nomatch.

%% @doc Split a URL path by `/`, discarding empty segments from leading slash.
-spec split_path(binary()) -> [binary()].
split_path(<<"/", Rest/binary>>) ->
    split_path(Rest);
split_path(Path) ->
    binary:split(Path, <<"/">>, [global, trim_all]).

%%====================================================================
%% Helpers
%%====================================================================

%% @doc Extract named bindings from a matched path.
%%
%% Given `Pattern = <<"/items/{id}/tags/{tag}">>` and
%% `Path = <<"/items/42/tags/colour">>`, returns
%% `#{<<"id">> => <<"42">>, <<"tag">> => <<"colour">>}`.
-spec extract_bindings(path_pattern(), binary()) -> bindings().
extract_bindings(Pattern, Path) ->
    case match_path(Pattern, Path) of
        {ok, Bindings} -> Bindings;
        nomatch        -> #{}
    end.

File: runtime-elixir/server/smithy_router.ex
defmodule SmithyRouter do
  @moduledoc """
  Request routing for Smithy-generated Elixir servers.

  A generated dispatcher module builds a route table from its service model
  and delegates to `SmithyRouter.dispatch/4` (or `dispatch/5` with an explicit
  table).  Each route maps an `{method, path_pattern}` pair to a
  `{module, function}` handler.

  ## Route table format

  ```elixir
  [
    {"GET",    "/items",      MyHandler, :list_items},
    {"POST",   "/items",      MyHandler, :create_item},
    {"GET",    "/items/{id}", MyHandler, :get_item},
    {"DELETE", "/items/{id}", MyHandler, :delete_item},
  ]
  ```

  Each entry is a 4-tuple `{method, path_pattern, module, function}`.

  ## Path patterns

  Segments enclosed in `{}` are label captures:

      "/items/{id}"          → matches "/items/42", binds "id" => "42"
      "/prefix/{+proxy}"     → greedy label, captures the rest of the path

  ## Handler protocol

  The handler `module.function/3` is called as:

      module.function(bindings, headers, body)

  where `bindings` is a `%{String.t() => String.t()}` map.

  It must return one of:

      {:ok, output}          → 200 JSON response
      {:ok, status, body}    → custom status
      {:error, reason}       → forwarded to SmithyServer.error_response/2
      :not_found             → 404
  """

  @type method       :: String.t()
  @type path_pattern :: String.t()
  @type route        :: {method(), path_pattern(), module(), atom()}
  @type routes       :: [route()]
  @type bindings     :: %{String.t() => String.t()}
  @type headers      :: map() | [{String.t(), String.t()}]
  @type body         :: binary()

  # ---------------------------------------------------------------------------
  # Public API
  # ---------------------------------------------------------------------------

  @doc """
  Dispatch a request using an explicit route table.

  Iterates routes in order and stops at the first match.
  Calls the matched handler with `(bindings, headers, body)`.

  Returns the handler's result, or `:not_found` when nothing matches.
  """
  @spec dispatch(routes(), method(), String.t(), headers(), body()) :: term()
  def dispatch(routes, method, path, headers, body) do
    case match(routes, method, path) do
      {:ok, bindings, mod, fun} -> apply(mod, fun, [bindings, headers, body])
      :nomatch                  -> :not_found
    end
  end

  @doc """
  Find the first matching route for `method` and `path`.

  Returns `{:ok, bindings, module, function}` or `:nomatch`.
  """
  @spec match(routes(), method(), String.t()) ::
          {:ok, bindings(), module(), atom()} | :nomatch
  def match([], _method, _path), do: :nomatch

  def match([{route_method, pattern, mod, fun} | rest], method, path) do
    if route_method == method or route_method == "*" do
      case match_path(pattern, path) do
        {:ok, bindings} -> {:ok, bindings, mod, fun}
        :nomatch        -> match(rest, method, path)
      end
    else
      match(rest, method, path)
    end
  end

  # ---------------------------------------------------------------------------
  # Path matching
  # ---------------------------------------------------------------------------

  @doc """
  Test whether `path` matches `pattern`, returning label bindings.

  ## Examples

      iex> SmithyRouter.match_path("/items/{id}", "/items/42")
      {:ok, %{"id" => "42"}}

      iex> SmithyRouter.match_path("/items/{id}", "/other/42")
      :nomatch
  """
  @spec match_path(path_pattern(), String.t()) :: {:ok, bindings()} | :nomatch
  def match_path(pattern, path) do
    pattern_segs = split_path(pattern)
    path_segs    = split_path(path)
    match_segments(pattern_segs, path_segs, %{})
  end

  @doc """
  Extract named bindings from a matched path.

  Returns an empty map when the pattern does not match.
  """
  @spec extract_bindings(path_pattern(), String.t()) :: bindings()
  def extract_bindings(pattern, path) do
    case match_path(pattern, path) do
      {:ok, bindings} -> bindings
      :nomatch        -> %{}
    end
  end

  @doc """
  Split a URL path on `/`, discarding leading empty segments.

  ## Examples

      iex> SmithyRouter.split_path("/items/42")
      ["items", "42"]
  """
  @spec split_path(String.t()) :: [String.t()]
  def split_path(path) do
    path
    |> String.trim_leading("/")
    |> String.split("/", trim: true)
  end

  # ---------------------------------------------------------------------------
  # Private helpers
  # ---------------------------------------------------------------------------

  defp match_segments([], [], bindings), do: {:ok, bindings}

  # Greedy label {+name} — consumes all remaining path segments
  defp match_segments(["{+" <> rest | _pat_rest], path_segs, bindings) do
    label = String.trim_trailing(rest, "}")
    value = Enum.join(path_segs, "/")
    {:ok, Map.put(bindings, label, value)}
  end

  # Regular label {name} — matches exactly one segment
  defp match_segments(["{" <> rest | pat_rest], [seg | path_rest], bindings) do
    label = String.trim_trailing(rest, "}")
    match_segments(pat_rest, path_rest, Map.put(bindings, label, seg))
  end

  # Literal segment — must match exactly
  defp match_segments([literal | pat_rest], [literal | path_rest], bindings) do
    match_segments(pat_rest, path_rest, bindings)
  end

  defp match_segments(_, _, _), do: :nomatch
end
